libdnf5/base/goal.cpp

```diff
@@ -2206,6 +2412,138 @@ GoalProblem Goal::Impl::add_reason_change_to_goal(
+
+    using Action = transaction::TransactionItemAction;
+    using Reason = transaction::TransactionItemReason;
+    const std::unordered_map<Action, Action> REVERT_ACTION = {
+        {Action::INSTALL, Action::REMOVE},
+        {Action::UPGRADE, Action::REPLACED},
+        {Action::DOWNGRADE, Action::REPLACED},
+        {Action::REINSTALL, Action::REINSTALL},
+        {Action::REMOVE, Action::INSTALL},
+        {Action::REPLACED, Action::INSTALL},
+        {Action::REASON_CHANGE, Action::REASON_CHANGE},
+    };
+    transaction::TransactionReplay replay;
+    auto history = base->get_transaction_history();
+
+    auto & [reverting_transactions, settings] = *revert_transactions;
+    //TODO(amatej): Implement merging of transactions and merge the vector
+    //              instead of taking the first one.
+    auto & reverting_transaction = reverting_transactions[0];
+
+    for (const auto & pkg : reverting_transaction.get_packages()) {
+        transaction::PackageReplay package_replay;
+        package_replay.nevra = libdnf5::rpm::to_nevra_string(pkg);
+        auto reverted_action = REVERT_ACTION.find(pkg.get_action());
+        libdnf_assert(
+            reverted_action != REVERT_ACTION.end(),
+            "Cannot revert action: \"{}\"",
+            transaction_item_action_to_string(pkg.get_action()));
+        package_replay.action = reverted_action->second;
+
+        // We cannot tell the previous reason if the action is REASON_CHANGE it could have been anything.
+        // For reverted action INSTALL and reason CLEAN the previous reason could have been either DEPENDENCY or WEAK DEPENDENCY
+        // to pick the right one we have to look into history.
+        if ((package_replay.action == Action::REASON_CHANGE) ||
+            (package_replay.action == Action::INSTALL && pkg.get_reason() == Reason::CLEAN)) {
+            auto nevras = rpm::Nevra::parse(pkg.to_string(), {rpm::Nevra::Form::NEVRA});
+            libdnf_assert(
+                nevras.size() > 0, "Cannot parse rpm nevra \"{}\" while reverting transaction.", pkg.to_string());
+            // We look up the reason based on only name and arch, this means we could find a different
+            // version of installonly package however we store only one reason for ALL versions of
+            // installonly packages so it doesn't matter.
+            package_replay.reason = history->transaction_item_reason_at(
+                nevras[0].get_name(), nevras[0].get_arch(), reverting_transaction.get_id() - 1);
+        } else if (
+            package_replay.action == Action::REMOVE &&
+            (pkg.get_reason() == Reason::DEPENDENCY || pkg.get_reason() == Reason::WEAK_DEPENDENCY)) {
+            package_replay.reason = Reason::CLEAN;
+        } else {
+            package_replay.reason = pkg.get_reason();
+        }
+
+        // Don't fill in @System repo_id because it's not possible to perform forward actions from it
+        if (pkg.get_repoid() != "@System") {
```

Comment:
I think that setting `repoid` is not right for undo operation or it must be performed differently.

Think about case reverting:
upgrade a-2-1.noarch  repository updates
`replaced a-1-1.noarch @system`

It is transformed to operation
remove a-2-1.noarch  repository to_repo_id updates
`install a-1-1.noarch o_repo_id @system`

For remove step it is ignored. For install step it is not not set. It means the code make no differences.

I see two options:
 - store installed from in history DB and use that for undo operation
 - Ignore repo for `undo` operation to allow to use optimizations like local repository. Personally I prefer this option for undo operation. But storing installed from remove steps makes also some volue, but without impact on `undo` command.

Reply:
I see, ok, I will drop the condition and ignore the repo.